### PR TITLE
AbstractReferenceCountedByteBuf can save using a volatile set on constructor

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

AbstractReferenceCountedByteBuf can be made faster by ensure proper safe initialization vs volatile set on construction

Modification:

Using Unsafe and plain store instead of volatile set

Result:

Faster AbstractReferenceCountedByteBuf allocation

### CODE REVIEW

**Comment by franz1981** [comment]
Still have to run benchs (that's why it's a draft) and need adding few checks to be sure the fence method is present on Unsafe.
This should benefit unpooled (array) buffer allocations, mostly

**Comment by franz1981** [comment]
it's similar to https://github.com/netty/netty/pull/12286

**Comment by franz1981** [comment]
Results with `delay == 0` on my machine are
```
4.1:
Benchmark                                                          (delay)  Mode  Cnt   Score   Error  Units
AbstractReferenceCountedByteBufBenchmark.createUseAndRelease             0  avgt    8  54.772 ± 0.445  ns/op
```
this PR:
```
Benchmark                                                          (delay)  Mode  Cnt   Score   Error  Units
AbstractReferenceCountedByteBufBenchmark.createUseAndRelease             0  avgt    8  47.956 ± 0.515  ns/op
```
There's an improvement, but not a huge one

**Comment by franz1981** [comment]
Relevant assembly parts are...
4.1:
```
  0.03%  │  0x00007f704d2aab23: movl   $0x2,0x20(%rbp)
  0.03%  │  0x00007f704d2aab2a: lock addl $0x0,(%rsp)     ;*putfield refCnt
         │                                                ; - io.netty.buffer.AbstractReferenceCountedByteBuf::&lt;init&gt;@12 (line 46)
         │                                                ; - io.netty.buffer.UnpooledHeapByteBuf::&lt;init&gt;@2 (line 51)
         │                                                ; - io.netty.buffer.UnpooledUnsafeHeapByteBuf::&lt;init&gt;@4 (line 34)
         │                                                ; - io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf::&lt;init&gt;@4 (line 139)
         │                                                ; - io.netty.buffer.UnpooledByteBufAllocator::newHeapBuffer@13 (line 82)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@21 (line 169)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@4 (line 160)
         │                                                ; - io.netty.buffer.Unpooled::buffer@4 (line 119)
         │                                                ; - io.netty.buffer.AbstractReferenceCountedByteBufBenchmark::createUseAndRelease@1 (line 66)
 15.16%  │  0x00007f704d2aab2f: movl   $0xe3932e95,0x24(%rbp)  ;   {oop(a &apos;io/netty/buffer/UnpooledByteBufAllocator&apos;)}
```

Hence, using storeFence translate in no-op (if we're lucky with inlining it till the Unsafe::storeFence intrinsic, and there's no associated cost, while on 4.1 there is a full barrier after setting the refCnt field with its initial value ie `0x2`, in the form of a `lock xadd` which cost is later shown ie `15.16%`, appeared later due to skidding

**Comment by franz1981** [comment]
IMO it still worth to be considered wdyt @normanmaurer ?
It looks similar to the optimization made by Nitsan years ago re using plain load on refCnt vs using atomic fences

File: buffer/src/main/java/io/netty/buffer/AbstractReferenceCountedByteBuf.java [hunk 1]
```diff
@@ -43,10 +44,16 @@ protected long unsafeOffset() {
 
     // Value might not equal "real" reference count, all access should be via the updater
     @SuppressWarnings({"unused", "FieldMayBeFinal"})
-    private volatile int refCnt = updater.initialValue();
+    private volatile int refCnt;
 
     protected AbstractReferenceCountedByteBuf(int maxCapacity) {
         super(maxCapacity);
+        if (!PlatformDependent.hasUnsafe()) {
```
[the remark this answers is not in the document]
Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/AbstractReferenceCountedByteBuf.java:51]: I think to move everything in a better named method too, but I'm curious if any can run some experiment on the existing allocations; I've a gut feeling (hopefully not wrong) that it has a larger benefit that expected given that it happen on hot path

**Comment by franz1981** [comment]
@normanmaurer I've implemented a specific PlatformDependent method to "safe construct put int", but I've been unlucky that it has hit the inlining depth limit of JDK 8 and *not* inlined, see
```
         │  0x00007f6edd2a90a0: mov    $0x2,%ecx
  0.02%  │  0x00007f6edd2a90a5: xchg   %ax,%ax
         │  0x00007f6edd2a90a7: callq  0x00007f6edd046260  ; OopMap{rbp=Oop [0]=Oop [8]=Oop [16]=Oop [24]=Oop off=204}
         │                                                ;*invokestatic safeConstructPutInt
         │                                                ; - io.netty.util.internal.PlatformDependent::safeConstructPutInt@3 (line 508)
         │                                                ; - io.netty.util.internal.ReferenceCountUpdater::setInitialValue@34 (line 67)
         │                                                ; - io.netty.buffer.AbstractReferenceCountedByteBuf::&lt;init&gt;@9 (line 50)
         │                                                ; - io.netty.buffer.UnpooledHeapByteBuf::&lt;init&gt;@2 (line 51)
         │                                                ; - io.netty.buffer.UnpooledUnsafeHeapByteBuf::&lt;init&gt;@4 (line 34)
         │                                                ; - io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf::&lt;init&gt;@4 (line 139)
         │                                                ; - io.netty.buffer.UnpooledByteBufAllocator::newHeapBuffer@13 (line 83)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@21 (line 169)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@4 (line 160)
         │                                                ; - io.netty.buffer.Unpooled::buffer@4 (line 119)
         │                                                ; - io.netty.buffer.AbstractReferenceCountedByteBufBenchmark::createUseAndRelease@1 (line 66)
         │                                                ;   {static_call}
         │  0x00007f6edd2a90ac: movl   $0xe393267d,0x24(%rbp)  ;   {oop(a &apos;io/netty/buffer/UnpooledByteBufAllocator&apos;)}
```
This is one of the drawback of using wrapped `Unsafe` in the form of `PlatformDependent` methods; they won't benefit from be intrinsified (-> that's the reason why on JCTools we directly use Unsafe) :(

And indeed the performance are now not as brilliant as they should be ie
```
Benchmark                                                          (delay)  Mode  Cnt   Score   Error  Units
AbstractReferenceCountedByteBufBenchmark.createUseAndRelease             0  avgt   10  50.324 ± 1.069  ns/op
```
Still better then the existing one, but will now pay the `callq` cost: If I run the same benchmark with `-XX:MaxInlineLevel=15`:
```
Benchmark                                                          (delay)  Mode  Cnt   Score   Error  Units
AbstractReferenceCountedByteBufBenchmark.createUseAndRelease             0  avgt   10  43.067 ± 0.868  ns/op
```
That seems a good improvement (as expected).

The assembly has changed too:
```
  0.02%  │  0x00007f3bb52a58f5: movq   $0x2,0x20(%rax)    ;*new  ; - io.netty.buffer.UnpooledByteBufAllocator::newHeapBuffer@6 (line 83)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@21 (line 169)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@4 (line 160)
         │                                                ; - io.netty.buffer.Unpooled::buffer@4 (line 119)
         │                                                ; - io.netty.buffer.AbstractReferenceCountedByteBufBenchmark::createUseAndRelease@1 (line 66)
         │  0x00007f3bb52a58fd: mov    %rax,0x10(%rsp)    ;*invokevirtual storeFence
         │                                                ; - io.netty.util.internal.PlatformDependent0::safeConstructPutInt@18 (line 616)
         │                                                ; - io.netty.util.internal.PlatformDependent::safeConstructPutInt@3 (line 508)
         │                                                ; - io.netty.util.internal.ReferenceCountUpdater::setInitialValue@34 (line 67)
         │                                                ; - io.netty.buffer.AbstractReferenceCountedByteBuf::&lt;init&gt;@9 (line 50)
         │                                                ; - io.netty.buffer.UnpooledHeapByteBuf::&lt;init&gt;@2 (line 51)
         │                                                ; - io.netty.buffer.UnpooledUnsafeHeapByteBuf::&lt;init&gt;@4 (line 34)
         │                                                ; - io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf::&lt;init&gt;@4 (line 139)
         │                                                ; - io.netty.buffer.UnpooledByteBufAllocator::newHeapBuffer@13 (line 83)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@21 (line 169)
         │                                                ; - io.netty.buffer.AbstractByteBufAllocator::heapBuffer@4 (line 160)
         │                                                ; - io.netty.buffer.Unpooled::buffer@4 (line 119)
         │                                                ; - io.netty.buffer.AbstractReferenceCountedByteBufBenchmark::createUseAndRelease@1 (line 66)
  0.89%  │  0x00007f3bb52a5902: movl   $0xe3880a40,0x24(%rax)  ;   {oop(a &apos;io/netty/buffer/UnpooledByteBufAllocator&apos;)}
  ```
The expected `movl` has been replaced by a `movq`, no idea why, given that the intiial value is an int field (mov*l*), not a quad mov*q*) and there's a wrong attribution to `storeFence`, that should be a no-op, but overall the improvement looks great.

In short: although a wild (and unconfigured) JMH benchmark shows just a small improvement (because of unlucky inlining depth of the newly platform dependent method), the improvement in the best case is larger.

**Comment by franz1981** [comment]
Re the inlining thing:

image

it shows that everything is getting inlined (as expected) with max inlining level 15 
while it's not with jdk 8 default max inlining level (9):
image

**Comment by franz1981** [comment]
PTAL @normanmaurer

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
@chrisvest @njhill PTAL

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
Thanks!
